[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)

laurencer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61980

Differential Revision: [D29917766](https://our.internmc.facebook.com/intern/diff/D29917766)

laurencer committed:
[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)

[ghstack-poisoned]

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"




[ghstack-poisoned]

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"




[ghstack-poisoned]

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
+      // elements. After processing of those elements is done, we move
+      // to the next block (the else block if it has nodes) from the
+      // `move_up` call.
+      auto* then_block = node->blocks().at(0);
```
this block i think recreates some of the logic that exists elsewhere. could we set current just equal to the beginning of the then block and if then_block is empty call move_next on it ?

laurencer replied:
I don't think this would work - if then doesn't have any nodes it would return the equivalent of a `nullptr` so the `move_next` function couldn't get a reference to the parent to `move_up`?

eellison replied:
Sorry, why can't we set `current_ = node->blocks().at(0)->nodes()->begin(); move_next(current_)` ? i still think that would work

laurencer replied:
I updated this in the PR. Let me know if the new approach works for you.

eellison commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
-        current_ = block_stack_.back();
-        block_stack_.pop_back();
+    } else if (node->kind() == prim::Loop || node->kind() == prim::With) {
+      auto* body_block = node->blocks().at(0);
```
would it work to set current_ to `node->blocks()->param_node()` and then call move_next() ? same question would apply for prim::If handling within this function as well

laurencer replied:
`param_node` doesn't have a next node. Per the docs in `OVERVIEW.md`

> The prim::Param node represents the inputs to the Block and does not have a prev() or next() Node.

eellison reviewed (commented):
Nice, looking good! A few comments

laurencer commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
+
+    // Check if we've reached the top of the graph.
+    if (parent_block == nullptr) {
+      current_ = graph_node_list_iterator();
```
current_ can't be set to `nullptr` directly. Initializing to `graph_node_list_iterator` initializes it to a nullptr.

laurencer commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
+    auto parent_block = from->owningBlock();
+
+    // Check if we've reached the top of the graph.
+    if (parent_block == nullptr) {
```
This was a check just for safety. I wasn't completely sure if a `Node` could be constructed without an owning block (e.g. if a graph can contain just the `Node` without a `Block`) I could change to an assert, remove, or leave as-is?

eellison replied:
Yea, if we're going to leave it in change it to an ASSERT please. Either that or remove it.

laurencer commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
+      // we're at the top of the graph and need to just move to the
+      // next node if possible.
+      current_ = parent_block->nodes().begin();
+      while (*current_ != from) {
```
Ah gotcha didn't realize what the semantics of next/prev were. I'll update the diff.

laurencer commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
-          block_stack_.emplace_back(current_.first, current_.second);
+  // Moves to the next node in the graph into children if it can.
+  void move_into() {
+    if (*current_ == nullptr) {
```
This is more for safety/defensive programming and to prevent issues for future refactorings. Basically if we are in a position where we can't move further we shouldn't.

eellison replied:
could we make this an ASSERTION instead ? should it be an error if this fires ?

laurencer replied:
Ack will do

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"




[ghstack-poisoned]

laurencer commented:
Updated the PR to address comments from @eellison

eellison commented on torch/csrc/jit/runtime/graph_iterator.h:
```diff
@@ -5,113 +5,196 @@ namespace jit {
+    // need to find the specific position of the owning node
+    // in case we need to go next or up.
+    auto owning_node_block = owning_node->owningBlock();
+    if (owning_node == nullptr) {
```
*invariants that are maintained by developers

eellison reviewed (approved):
Great job 😍 😍 😍  And very well tested to boot!

laurencer commented:
@laurencer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29917766).

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"


Differential Revision: [D29917766](https://our.internmc.facebook.com/intern/diff/D29917766)

[ghstack-poisoned]

laurencer commented:
@laurencer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29917766).

laurencer committed:
Update on "[jit] Rewrote DFS graph iterator to remove unnecessary local state (#61326)"


Differential Revision: [D29917766](https://our.internmc.facebook.com/intern/diff/D29917766)

[ghstack-poisoned]

laurencer commented:
@laurencer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29917766).
